### PR TITLE
Chapter 10 updating users

### PR DESCRIPTION
**Pull request by Arizar98**

### CODE REVIEW

**Comment by Arizar98** [comment]
Ready

File: Gemfile [hunk 1]
```diff
@@ -15,6 +15,9 @@ gem "bcrypt", "3.1.13"
 gem "bootsnap", ">= 1.4.2", require: false
 gem "rails-i18n"
 gem "jquery-rails"
+gem "faker", "1.7.3"
+gem "will_paginate", "3.1.7"
+gem "bootstrap-will_paginate", "1.0.0"
```
Comment (saobv-2223) [inline Gemfile]: using gem kaminari to paginate instead of gem will_paginate

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,4 +1,17 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
```
```suggestion
  before_action :logged_in_user, except: %i(new create show)
```

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,4 +1,17 @@
...
+  before_action :correct_user, only: [:edit, :update]
+
+  def index
+    @users = User.paginate(page: params[:page])
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -22,9 +35,40 @@ def create
     end
   end
 
+  def edit
+    @user = User.find_by id: params[:id]
+  end
+
+  def update
+    @user = User.find_by(id: params[:id])
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -22,9 +35,40 @@ def create
...
+    if @user.update_attributes(user_params)
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -22,9 +35,40 @@ def create
...
+    unless logged_in?
+      store_location
+      flash[:danger] = t "please_login"
+      redirect_to login_url
+    end
+  end
+
+  # Confirms the correct user.
+  def correct_user
+    @user = User.find_by(id: params[:id])
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -48,4 +48,20 @@ def log_out
...
+
+  # Returns true if the given user is the current user.
+  def current_user?(user)
+    user == current_user
+  end
+
+  # Redirects to stored location (or to the default).
+  def redirect_back_or(default)
+    redirect_to(session[:forwarding_url] || default)
+    session.delete(:forwarding_url)
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb:60]: remove ()

File: app/views/users/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<%= form_for(@user) do |f| %>
```
Comment (saobv-2223) [inline app/views/users/_form.html.erb]: remove ()

File: app/views/users/edit.html.erb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<% provide(:title, "Edit user") %>
```
Comment (saobv-2223) [inline app/views/users/edit.html.erb]: remove ()

File: app/views/users/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+<% provide(:button_text, "Save changes") %>
```
Comment (saobv-2223) [inline app/views/users/edit.html.erb]: remove ()

File: app/views/users/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<% provide(:title, "All users") %>
```
Comment (saobv-2223) [inline app/views/users/index.html.erb]: remove ()

File: app/views/users/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+<h1>All users</h1>
+
+<%= will_paginate %>
```
Comment (saobv-2223) [inline app/views/users/index.html.erb]: using gem kaminari to paginate instead of gem will_paginate

**Comment by Arizar98** [comment]
Ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**